zpay32/invoice_internal_test.go

```diff
@@ -230,13 +296,510 @@ func TestParseTimestamp(t *testing.T) {
 	for i, test := range tests {
 		time, err := parseTimestamp(test.data)
 		if (err == nil) != test.valid {
-			t.Errorf("Data decoding test %d failed: %v", i, err)
+			t.Errorf("timestamp decoding test %d failed: %v", i, err)
 			return
 		}
 		if test.valid && time != test.result {
-			t.Errorf("Timestamp decoding test %d failed: expected "+
-				"timestamp %d, got %d", i, test.result, time)
+			t.Errorf("%d) failed decoding timestamp: "+
+				"expected %d, got %d",
+				i, test.result, time)
+			return
+		}
+	}
+}
+
+// TestParseFieldDataLength checks that the 16 bit length is properly parsed.
+func TestParseFieldDataLength(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		data   []byte
+		valid  bool
+		result uint16
+	}{
+		{
+			data:  []byte{},
+			valid: false, // empty data
+		},
+		{
+			data:  []byte{0x0},
+			valid: false, // data too short
+		},
+		{
+			data:  []byte{0x0, 0x0, 0x0},
+			valid: false, // data too long
+		},
+		{
+			// The first byte is <= 3 bits long.
+			data:   []byte{0x1, 0x2},
+			valid:  true,
+			result: 34,
+		},
+		{
+			// The first byte is > 3 bits long.
+			data:   []byte{0xa, 0x0},
+			valid:  true,
+			result: 320,
+		},
+	}
+
+	for i, test := range tests {
+		length, err := parseFieldDataLength(test.data)
+		if (err == nil) != test.valid {
+			t.Errorf("field data length decoding test %d failed: %v", i, err)
+			return
+		}
+		if test.valid && length != test.result {
+			t.Errorf("%d) failed decoding field data length: "+
+				"expected %d, got %d",
+				i, test.result, length)
+			return
+		}
+	}
+}
+
+// TestParsePaymentHash checks that the payment hash is properly parsed.
+func TestParsePaymentHash(t *testing.T) {
+	t.Parallel()
+
+	var testPaymentHash [32]byte
+	copy(testPaymentHash[:], testPaymentHashSlice[:])
+
+	tests := []struct {
+		data   []byte
+		valid  bool
+		result *[32]byte
+	}{
+		{
+			data:  []byte{},
+			valid: false, // empty data
+		},
+		{
+			data:  []byte{0x00, 0x00, 0x00, 0x00, 0x00, 0x00},
+			valid: false, // data too short, not 52 bytes
+		},
+		{
+			data:   []byte{0x0, 0x0, 0x0, 0x10, 0x4, 0x0, 0x18, 0x4, 0x0, 0x14, 0x3, 0x0, 0xe, 0x2, 0x0, 0x9, 0x0, 0x0, 0x0, 0x10, 0x4, 0x0, 0x18, 0x4, 0x0, 0x14, 0x3, 0x0, 0xe, 0x2, 0x0, 0x9, 0x0, 0x0, 0x0, 0x10, 0x4, 0x0, 0x18, 0x4, 0x0, 0x14, 0x3, 0x0, 0xe, 0x2, 0x0, 0x9, 0x0, 0x4, 0x1, 0x0},
```

Comment:
instead of hardcoding the data slice here, could you extract it from the expected result? You should be able to use the `ConvertBits` method on the payment hash to get this data.